brainiak/utils/fmrisim.py

```diff
@@ -804,10 +826,13 @@ def convolve_hrf(stimfunction,
         columns in this array.
 
     """
+    # How will stimfunction be resized
+    stride = int(temporal_resolution * tr_duration)
+    duration = int(stimfunction.shape[0] / stride)
 
     # Generate the hrf to use in the convolution
     if hrf_type == 'double_gamma':
-        hrf = _double_gamma_hrf(temporal_resolution=temporal_resolution)
+        hrf = _double_gamma_hrf(temporal_resolution=1 / tr_duration)
```

Comment:
Although down-sampling stim_function helps speed up the convolution, I worry that this defeats the original purpose of building a smooth double_gamma_hrf. Consider a case of TR=2s, a stimulus of condition A starts at 1.5s and lasts for 0.5s, and a stimulus of condition B starts at 2s and lasts for 0.5s. I think the original implementation would make the design matrix of the two conditions appear as down-sampled version of two HRF with onsets 0.5 sec apart (they would have slightly different shapes, but very close to each other). But the new implementation might make them exactly the same shape, but 2 sec apart. Is this true? If so, maybe the old implementation is more desirable. You may set the default temporal resolution to some lower number such as 100 or 50 to achieve some speed-up with the old implementation. (most often the displays in scanners have refresh rate of just 60 Hz)

Reply:
Fantastic point. I had been testing this in cases where the event duration was greater than the TR (and so this isn't a problem). But as you point out when the events are shorter than a TR then you get this exact problem. I have shifted the downsampling step back to after convolution. I also changed the default temporal_resolution: it was 1000 so that it was measuring milliseconds, but you are right >100Hz is excessive